File: HW5/151044058_CSE241_HW5/PolygonDyn.cpp
#include "Circle.h"
#include "Triangle.h"
#include "Rectangle.h"
#include "Polygon.h"
#include "Shape.h"
#include "PolygonDyn.h"

namespace HW5_151044058
{
	PolygonDyn :: PolygonDyn()
	{
		//Empty
	}
	PolygonDyn :: PolygonDyn(const PolygonDyn& pol1) // Copy constructor
	{
		int i;
		size = pol1.size;
		ptr = new Point2D[size];
		for(i=0; i<size; ++i)
		{
			ptr[i] = pol1.ptr[i];
		}
	}

	PolygonDyn :: ~PolygonDyn()// Destructor
	{
		delete [] ptr;
		ptr = nullptr;
	}

	PolygonDyn& PolygonDyn :: operator=(const PolygonDyn& pol1)// Assignment operator
	{
		if(size != pol1.size)
		{
			size = pol1.size;
			delete [] ptr;
			ptr = new Point2D[size];
		}
		for(int i=0; i<size; ++i)
		{
			ptr[i] = pol1.ptr[i];
		}
		return *this;
	} 

	PolygonDyn :: PolygonDyn(const Rectangle &rec) //	conversion rectangle
	{
		
		int temp_size = 4;
		size = temp_size;
		ptr = new Point2D[size]; //Yer ayirma
		Point2D p1(rec.getCor_x(),rec.getCor_y());
		ptr[0] = p1;
		Point2D p2(rec.getCor_x() + rec.getWidth() ,rec.getCor_y());
		ptr[1] = p2;
		Point2D p3(rec.getCor_x() + rec.getWidth(),rec.getCor_y() + rec.getHeight());
		ptr[2] = p3;
		Point2D p4(rec.getCor_x(),rec.getCor_y() + rec.getHeight());
		ptr[3] = p4;

	}

	PolygonDyn :: PolygonDyn(const Triangle &tri) //	conversion triangle
	{
		int temp_size = 3;
		size = temp_size;
		ptr = new Point2D[size];	//Yer ayirma

		Point2D p1(tri.getCor_x1(),tri.getCor_y1());
		ptr[0] = p1;
		Point2D p2(tri.getCor_x2(),tri.getCor_y2());
		ptr[1] = p2;
		Point2D p3(tri.getCor_x3(),tri.getCor_y3());
		ptr[2] = p3;
	}

	PolygonDyn :: PolygonDyn(const Circle &cir)  //conversion circle
	{
		const double PI = M_PI;
		int i,temp_size = 100;
		size = temp_size;
		ptr = new Point2D[size]; //Yer ayirma

		//Circle donusum formulu
		for(i = 1; i <= size; ++i)
		{
			Point2D p1(cir.getCenter_x()+(cir.getRadius()*sin(((2.0*PI)/100.0)*i)), cir.getCenter_y()+(cir.getRadius()*cos(((2.0*PI)/100.0)*i)));
			ptr[i-1] = p1;
		}
	}

	void PolygonDyn :: draw_Shape(ostream& file)
	{
		int i;

		file << "<polygon points=\"";
		for(i = 0; i < this->size; ++i)
		{
			file << this->ptr[i].getKoor_x()<<" "<< this->ptr[i].getKoor_y()<<" ";	 
		}
		file<< "\" fill=\"blue\" stroke=\"white\" />\n" ;
	}
	
	PolygonDyn :: Point2D& PolygonDyn::operator[](int index) // idex operatoru
	{
		if(index >= size)
			index = size - 1;

		return ptr[index];

	}
	int PolygonDyn :: getSize()const
	{
		return size;
	}
	void PolygonDyn :: setSize(int sizeValue)
	{
		size = sizeValue;
	}
}

File: HW7/151044058_CSE241_HW7/PolygonDyn.java
import java.awt.Color;
import java.awt.Graphics;

/**
 * 
 * @author Ali Haydar Kurban
 *
 */
public class PolygonDyn extends Polygon 
{
	/**
	 * 
	 */
	private static final long serialVersionUID = 1L;
	private Point2D [] point2D_Array;
	private	int size; // dynamic array size
	private double alan;
	private double cevre;
	

	@Override
	public void draw(Graphics g)
	{
		int [] array1 = new int [point2D_Array.length];
		int [] array2 = new int [point2D_Array.length];
		for(int i = 0; i < point2D_Array.length; ++i)
		{
			array1[i] = (int)point2D_Array[i].getKoor_x();
			array2[i] = (int)point2D_Array[i].getKoor_y();
		}
		g.setColor(Color.BLACK);
		g.fillPolygon(array1, array2,point2D_Array.length);
	}
	

	@Override
	public void paintComponent(Graphics g)
	{
		super.paintComponent(g);
		draw(g);
	}
	
	/**
	 * 
	 */
	public PolygonDyn()
	{
		//Empty
	}
	
	/**
	 * 
	 * @param rec objesini PolygonDyn ye donusturme constructorudur
	 */
	public PolygonDyn(Rectangle rec)
	{
		int temp_size = 4;
		size = temp_size;
		point2D_Array = new Point2D[size];
		
		Point2D p1 = new Point2D(rec.getCor_x(),rec.getCor_y());
		point2D_Array[0] = p1;
		
		Point2D p2 = new Point2D(rec.getCor_x() + rec.getWidthh() ,rec.getCor_y());
		point2D_Array[1] = p2;
		
		Point2D p3 = new Point2D(rec.getCor_x() + rec.getWidthh(),rec.getCor_y() + rec.getHeightt());
		point2D_Array[2] = p3;
		
		Point2D p4 = new Point2D(rec.getCor_x(),rec.getCor_y() + rec.getHeightt());
		point2D_Array[3] = p4;
		
		alan = rec.area();
		cevre = rec.perimeter();
	}
	
	/**
	 * 
	 * @param tri objesini PolygonDyn ye donusturme constructorudur
	 */
	public PolygonDyn(Triangle tri)
	{
		int temp_size = 3;
		size = temp_size;
		point2D_Array = new Point2D[size];

		Point2D p1 = new Point2D(tri.getCor_x1(),tri.getCor_y1());
		point2D_Array[0] = p1;
		
		Point2D p2 = new Point2D(tri.getCor_x2(),tri.getCor_y2());
		point2D_Array[1] = p2;
		
		Point2D p3 = new Point2D(tri.getCor_x3(),tri.getCor_y3());
		point2D_Array[2] = p3;
	
		alan = tri.area();
		cevre = tri.perimeter();
	}
	
	
	/**
	 * 
	 * @param cir objesini PolygonDyn ye donusturme constructorudur
	 */
	public PolygonDyn(Circle cir)
	{
		int i;
		int temp_size = 100;
		size = temp_size;
		
		point2D_Array = new Point2D[size];

		for(i = 1; i <= size; ++i)
		{
			Point2D p1 = new Point2D(cir.getRadius()*2 + 2 * (cir.getRadius()* Math.sin(((2.0*Math.PI)/100.0)*i)), cir.getRadius()*2 + 2*(cir.getRadius()*Math.cos(((2.0*Math.PI)/100.0)*i)));
			point2D_Array[i-1] = p1;
		}
		alan = cir.area();
		cevre = cir.perimeter();
	}
	
	/**
	 * 
	 * @return size return eder
	 */
	public int getSizeeee()
	{
		return size;
	}
	
	/**
	 * 
	 * @param sizeValue size a set etmek icin kullanilir
	 */
	public void setSize(int sizeValue)
	{
		size = sizeValue;
	}

	/**
	 * Alan methodu
	 */
	public double area()
	{
		return alan;
	}
	
	/**
	 * Cevre methodu
	 */
	public double perimeter()
	{
		return cevre;
	}
	
	/**
	 * Koordinatlari bir arttirir
	 */
	@Override
	public Shape increment()
	{
		for(int i = 0; i < point2D_Array.length; ++ i)
		{
			point2D_Array[i].setKoor_x(point2D_Array[i].getKoor_x() + 1);
			point2D_Array[i].setKoor_y(point2D_Array[i].getKoor_y() + 1);
		}
		return this;
	}
	
	/**
	 * Koordinatlari bir azaltir
	 */
	@Override
	public Shape decrement()
	{
		for(int i = 0; i < point2D_Array.length; ++ i)
		{
			point2D_Array[i].setKoor_x(point2D_Array[i].getKoor_x() - 1);
			point2D_Array[i].setKoor_y(point2D_Array[i].getKoor_y() - 1);
		}
		return this;
	}
	
	/**
	 * Alanlarin kiyaslanmasina gore compareTo fonksiyonunun override hali
	 */
	@Override
	public int compareTo(Object arg0) 
	{
		if(((Shape) arg0).area() > this.area())
			return -1;
		if(((Shape) arg0).area() < this.area())
			return 1;
		return 0;
	}

}